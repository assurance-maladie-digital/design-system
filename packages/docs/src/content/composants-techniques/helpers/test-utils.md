---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
