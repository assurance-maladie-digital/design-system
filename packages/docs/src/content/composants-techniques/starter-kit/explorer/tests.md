Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
