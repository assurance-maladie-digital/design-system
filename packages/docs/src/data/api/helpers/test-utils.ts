import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
