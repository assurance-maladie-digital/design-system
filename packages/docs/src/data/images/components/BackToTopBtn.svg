<svg width="541" height="357" viewBox="0 0 541 357" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_147_105475)">
<rect width="541" height="357" fill="#E7ECF5"/>
<g filter="url(#filter0_dddddd_147_105475)">
<g clip-path="url(#clip1_147_105475)">
<rect width="780" height="56" transform="translate(62 60)" fill="#E2E7EE"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M92 94C95.3137 94 98 91.3137 98 88C98 84.6863 95.3137 82 92 82C88.6863 82 86 84.6863 86 88C86 91.3137 88.6863 94 92 94Z" fill="#ED6A5E"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M112 94C115.314 94 118 91.3137 118 88C118 84.6863 115.314 82 112 82C108.686 82 106 84.6863 106 88C106 91.3137 108.686 94 112 94Z" fill="#F4BF4F"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M132 94C135.314 94 138 91.3137 138 88C138 84.6863 135.314 82 132 82C128.686 82 126 84.6863 126 88C126 91.3137 128.686 94 132 94Z" fill="#61C554"/>
<rect x="62" y="116" width="780" height="344" fill="#F6F8FC"/>
</g>
</g>
<g clip-path="url(#clip2_147_105475)">
<path d="M376.328 317V306.536H379.864C380.589 306.536 381.24 306.632 381.816 306.824C382.403 307.016 382.861 307.341 383.192 307.8C383.533 308.248 383.704 308.861 383.704 309.64C383.704 310.44 383.512 311.085 383.128 311.576C382.755 312.067 382.253 312.419 381.624 312.632L384.12 317H382.04L379.784 312.904H378.184V317H376.328ZM378.184 311.432H379.672C380.387 311.432 380.931 311.283 381.304 310.984C381.688 310.675 381.88 310.227 381.88 309.64C381.88 309.043 381.688 308.627 381.304 308.392C380.931 308.147 380.387 308.024 379.672 308.024H378.184V311.432ZM388.881 317.192C388.155 317.192 387.499 317.032 386.913 316.712C386.326 316.381 385.862 315.912 385.521 315.304C385.179 314.685 385.008 313.944 385.008 313.08C385.008 312.227 385.179 311.491 385.521 310.872C385.873 310.253 386.326 309.779 386.881 309.448C387.435 309.117 388.017 308.952 388.625 308.952C389.339 308.952 389.937 309.112 390.417 309.432C390.897 309.741 391.259 310.179 391.505 310.744C391.75 311.299 391.873 311.944 391.873 312.68C391.873 313.064 391.846 313.363 391.793 313.576H386.801C386.886 314.269 387.137 314.808 387.553 315.192C387.969 315.576 388.491 315.768 389.121 315.768C389.462 315.768 389.777 315.72 390.065 315.624C390.363 315.517 390.657 315.373 390.945 315.192L391.569 316.344C391.195 316.589 390.779 316.792 390.321 316.952C389.862 317.112 389.382 317.192 388.881 317.192ZM386.785 312.328H390.273C390.273 311.72 390.139 311.245 389.873 310.904C389.606 310.552 389.206 310.376 388.673 310.376C388.214 310.376 387.803 310.547 387.441 310.888C387.089 311.219 386.87 311.699 386.785 312.328ZM396.282 317.192C395.375 317.192 394.73 316.931 394.346 316.408C393.972 315.875 393.786 315.187 393.786 314.344V310.6H392.666V309.224L393.882 309.144L394.106 307H395.642V309.144H397.642V310.6H395.642V314.344C395.642 315.261 396.01 315.72 396.746 315.72C396.884 315.72 397.028 315.704 397.178 315.672C397.327 315.629 397.455 315.587 397.562 315.544L397.882 316.904C397.668 316.979 397.423 317.043 397.146 317.096C396.879 317.16 396.591 317.192 396.282 317.192ZM402.301 317.192C401.65 317.192 401.042 317.032 400.477 316.712C399.912 316.381 399.453 315.912 399.101 315.304C398.749 314.685 398.573 313.944 398.573 313.08C398.573 312.205 398.749 311.464 399.101 310.856C399.453 310.237 399.912 309.768 400.477 309.448C401.042 309.117 401.65 308.952 402.301 308.952C402.962 308.952 403.576 309.117 404.141 309.448C404.706 309.768 405.165 310.237 405.517 310.856C405.869 311.464 406.045 312.205 406.045 313.08C406.045 313.944 405.869 314.685 405.517 315.304C405.165 315.912 404.706 316.381 404.141 316.712C403.576 317.032 402.962 317.192 402.301 317.192ZM402.301 315.688C402.877 315.688 403.33 315.448 403.661 314.968C403.992 314.488 404.157 313.859 404.157 313.08C404.157 312.291 403.992 311.656 403.661 311.176C403.33 310.696 402.877 310.456 402.301 310.456C401.725 310.456 401.272 310.696 400.941 311.176C400.621 311.656 400.461 312.291 400.461 313.08C400.461 313.859 400.621 314.488 400.941 314.968C401.272 315.448 401.725 315.688 402.301 315.688ZM410.242 317.192C409.399 317.192 408.786 316.92 408.402 316.376C408.018 315.832 407.826 315.064 407.826 314.072V309.144H409.666V313.832C409.666 314.483 409.762 314.941 409.954 315.208C410.146 315.475 410.455 315.608 410.882 315.608C411.223 315.608 411.522 315.523 411.778 315.352C412.045 315.181 412.327 314.904 412.626 314.52V309.144H414.466V317H412.962L412.818 315.848H412.77C412.429 316.253 412.055 316.579 411.65 316.824C411.245 317.069 410.775 317.192 410.242 317.192ZM416.837 317V309.144H418.357L418.485 310.536H418.549C418.826 310.024 419.162 309.635 419.557 309.368C419.951 309.091 420.357 308.952 420.773 308.952C421.146 308.952 421.445 309.005 421.669 309.112L421.349 310.712C421.21 310.669 421.082 310.637 420.965 310.616C420.847 310.595 420.703 310.584 420.533 310.584C420.223 310.584 419.898 310.707 419.557 310.952C419.215 311.187 418.922 311.603 418.677 312.2V317H416.837ZM429.449 317.192C428.723 317.192 428.067 317.032 427.481 316.712C426.894 316.381 426.43 315.912 426.089 315.304C425.747 314.685 425.577 313.944 425.577 313.08C425.577 312.227 425.747 311.491 426.089 310.872C426.441 310.253 426.894 309.779 427.449 309.448C428.003 309.117 428.585 308.952 429.193 308.952C429.907 308.952 430.505 309.112 430.985 309.432C431.465 309.741 431.827 310.179 432.073 310.744C432.318 311.299 432.441 311.944 432.441 312.68C432.441 313.064 432.414 313.363 432.361 313.576H427.369C427.454 314.269 427.705 314.808 428.121 315.192C428.537 315.576 429.059 315.768 429.689 315.768C430.03 315.768 430.345 315.72 430.633 315.624C430.931 315.517 431.225 315.373 431.513 315.192L432.137 316.344C431.763 316.589 431.347 316.792 430.889 316.952C430.43 317.112 429.95 317.192 429.449 317.192ZM427.353 312.328H430.841C430.841 311.72 430.707 311.245 430.441 310.904C430.174 310.552 429.774 310.376 429.241 310.376C428.782 310.376 428.371 310.547 428.009 310.888C427.657 311.219 427.438 311.699 427.353 312.328ZM434.238 317V309.144H435.758L435.886 310.2H435.95C436.302 309.859 436.686 309.565 437.102 309.32C437.518 309.075 437.993 308.952 438.526 308.952C439.369 308.952 439.982 309.224 440.366 309.768C440.75 310.312 440.942 311.08 440.942 312.072V317H439.102V312.312C439.102 311.661 439.006 311.203 438.814 310.936C438.622 310.669 438.307 310.536 437.87 310.536C437.529 310.536 437.225 310.621 436.958 310.792C436.702 310.952 436.409 311.192 436.078 311.512V317H434.238ZM446.474 317V305.704H448.314V308.616L448.25 310.136C448.581 309.827 448.949 309.555 449.354 309.32C449.76 309.075 450.229 308.952 450.762 308.952C451.605 308.952 452.218 309.224 452.602 309.768C452.986 310.312 453.178 311.08 453.178 312.072V317H451.338V312.312C451.338 311.661 451.242 311.203 451.05 310.936C450.858 310.669 450.544 310.536 450.106 310.536C449.765 310.536 449.461 310.621 449.194 310.792C448.938 310.952 448.645 311.192 448.314 311.512V317H446.474ZM457.34 317.192C456.657 317.192 456.097 316.979 455.66 316.552C455.233 316.125 455.02 315.571 455.02 314.888C455.02 314.045 455.388 313.395 456.124 312.936C456.86 312.467 458.033 312.147 459.644 311.976C459.633 311.56 459.521 311.203 459.308 310.904C459.105 310.595 458.737 310.44 458.204 310.44C457.82 310.44 457.441 310.515 457.068 310.664C456.705 310.813 456.348 310.995 455.996 311.208L455.324 309.976C455.761 309.699 456.252 309.459 456.796 309.256C457.351 309.053 457.937 308.952 458.556 308.952C459.537 308.952 460.268 309.245 460.748 309.832C461.239 310.408 461.484 311.245 461.484 312.344V317H459.964L459.836 316.136H459.772C459.42 316.435 459.041 316.685 458.636 316.888C458.241 317.091 457.809 317.192 457.34 317.192ZM457.932 315.752C458.252 315.752 458.54 315.677 458.796 315.528C459.063 315.368 459.345 315.155 459.644 314.888V313.128C458.577 313.267 457.836 313.475 457.42 313.752C457.004 314.019 456.796 314.349 456.796 314.744C456.796 315.096 456.903 315.352 457.116 315.512C457.329 315.672 457.601 315.752 457.932 315.752ZM466.062 317.192C465.219 317.192 464.606 316.92 464.222 316.376C463.838 315.832 463.646 315.064 463.646 314.072V309.144H465.486V313.832C465.486 314.483 465.582 314.941 465.774 315.208C465.966 315.475 466.275 315.608 466.702 315.608C467.043 315.608 467.342 315.523 467.598 315.352C467.865 315.181 468.147 314.904 468.446 314.52V309.144H470.286V317H468.782L468.638 315.848H468.59C468.249 316.253 467.875 316.579 467.47 316.824C467.065 317.069 466.595 317.192 466.062 317.192ZM475.425 317.192C474.518 317.192 473.873 316.931 473.489 316.408C473.115 315.875 472.929 315.187 472.929 314.344V310.6H471.809V309.224L473.025 309.144L473.249 307H474.785V309.144H476.785V310.6H474.785V314.344C474.785 315.261 475.153 315.72 475.889 315.72C476.027 315.72 476.171 315.704 476.321 315.672C476.47 315.629 476.598 315.587 476.705 315.544L477.025 316.904C476.811 316.979 476.566 317.043 476.289 317.096C476.022 317.16 475.734 317.192 475.425 317.192Z" fill="#0C419A"/>
<path d="M486.667 311L487.607 311.94L491.333 308.22V316.333H492.667V308.22L496.387 311.947L497.333 311L492 305.667L486.667 311Z" fill="#0C419A"/>
</g>
<rect x="359.5" y="293.5" width="156" height="35" rx="3.5" stroke="#0C419A"/>
<rect opacity="0.5" x="95" y="155" width="409" height="20" rx="10" fill="#CED9EB"/>
<rect opacity="0.5" x="95" y="196" width="409" height="20" rx="10" fill="#CED9EB"/>
<rect opacity="0.5" x="95" y="237" width="409" height="20" rx="10" fill="#CED9EB"/>
</g>
<defs>
<filter id="filter0_dddddd_147_105475" x="-18" y="60" width="940" height="580" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
<feFlood flood-opacity="0" result="BackgroundImageFix"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="2.76726"/>
<feGaussianBlur stdDeviation="1.1069"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0196802 0"/>
<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_147_105475"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="6.6501"/>
<feGaussianBlur stdDeviation="2.66004"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0282725 0"/>
<feBlend mode="normal" in2="effect1_dropShadow_147_105475" result="effect2_dropShadow_147_105475"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="12.5216"/>
<feGaussianBlur stdDeviation="5.00862"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.035 0"/>
<feBlend mode="normal" in2="effect2_dropShadow_147_105475" result="effect3_dropShadow_147_105475"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="22.3363"/>
<feGaussianBlur stdDeviation="8.93452"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0417275 0"/>
<feBlend mode="normal" in2="effect3_dropShadow_147_105475" result="effect4_dropShadow_147_105475"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="41.7776"/>
<feGaussianBlur stdDeviation="16.711"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0503198 0"/>
<feBlend mode="normal" in2="effect4_dropShadow_147_105475" result="effect5_dropShadow_147_105475"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="100"/>
<feGaussianBlur stdDeviation="40"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.07 0"/>
<feBlend mode="normal" in2="effect5_dropShadow_147_105475" result="effect6_dropShadow_147_105475"/>
<feBlend mode="normal" in="SourceGraphic" in2="effect6_dropShadow_147_105475" result="shape"/>
</filter>
<clipPath id="clip0_147_105475">
<rect width="541" height="357" fill="white"/>
</clipPath>
<clipPath id="clip1_147_105475">
<rect x="62" y="60" width="780" height="400" rx="16" fill="white"/>
</clipPath>
<clipPath id="clip2_147_105475">
<rect x="359" y="293" width="157" height="36" rx="4" fill="white"/>
</clipPath>
</defs>
</svg>
