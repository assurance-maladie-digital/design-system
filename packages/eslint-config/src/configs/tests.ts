import { type FlatESLintConfigItem } from 'eslint-define-config';
import { pluginVitest } from '../plugins';
import { GLOB_TESTS } from '../globs';
import { isInEditor } from '../env';

export const tests: FlatESLintConfigItem[] = [
	{
		plugins: {
			tests: pluginVitest,
		},
	},
	{
		files: GLOB_TESTS,
		rules: {
			'tests/consistent-test-it': [
				'error',
				{
					fn: 'it',
					withinDescribe: 'it',
				},
			],
			'tests/expect-expect': 'error',
			'tests/no-alias-methods': 'error',
			'tests/no-commented-out-tests': isInEditor ? 'off' : 'error',
			'tests/no-conditional-expect': 'error',
			'tests/no-disabled-tests': 'error',
			'tests/no-done-callback': 'error',
			'tests/no-duplicate-hooks': 'error',
			'tests/no-focused-tests': isInEditor ? 'off' : 'error',
			'tests/no-identical-title': 'error',
			'tests/no-interpolation-in-snapshots': 'error',
			'tests/no-standalone-expect': 'error',
			'tests/no-test-return-statement': 'error',
			'tests/prefer-comparison-matcher': 'error',
			'tests/prefer-equality-matcher': 'error',
			'tests/prefer-hooks-in-order': 'error',
			'tests/prefer-hooks-on-top': 'error',
			'tests/prefer-mock-promise-shorthand': 'error',
			'tests/prefer-spy-on': 'error',
			'tests/prefer-to-be-falsy': 'error',
			'tests/prefer-to-be-object': 'error',
			'tests/prefer-to-be-truthy': 'error',
			'tests/prefer-to-be': 'error',
			'tests/prefer-to-contain': 'error',
			'tests/prefer-to-have-length': 'error',
			'tests/prefer-todo': 'error',
			'tests/require-hook': 'error',
			'tests/require-top-level-describe': 'error',
			'tests/valid-describe-callback': 'error',
			'tests/valid-expect': 'error',
			'tests/valid-title': 'error',
		},
	},
];
