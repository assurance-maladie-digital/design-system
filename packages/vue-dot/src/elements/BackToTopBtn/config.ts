export const config = {
	btn: {
		fixed: true,
		outlined: true,
		color: 'primary',
		class: 'text-wrap px-0 px-md-4'
	},
	icon: {
		color: 'primary',
		small: true,
		class: 'ml-0 ml-md-1'
	}
};
