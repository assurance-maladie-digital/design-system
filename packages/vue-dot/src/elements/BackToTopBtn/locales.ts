export const locales = {
	label: 'Retour en haut'
};
