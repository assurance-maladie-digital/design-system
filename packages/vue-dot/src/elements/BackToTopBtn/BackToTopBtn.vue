<template>
	<VFadeTransition>
		<VBtn
			v-show="showBtn"
			v-scroll:[targetSelector]="onScroll"
			v-bind="{
				...options.btn,
				...$attrs
			}"
			:style="btnStyle"
			:min-width="minWidth"
			class="vd-back-to-top-btn"
			@click="scrollToTop"
			v-on="$listeners"
		>
			<span :class="labelClasses">
				<slot>
					{{ locales.label }}
				</slot>
			</span>

			<slot name="icon">
				<VIcon v-bind="options.icon">
					{{ topIcon }}
				</VIcon>
			</slot>
		</VBtn>
	</VFadeTransition>
</template>

<script lang="ts">
	import Vue from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { customizable } from '../../mixins/customizable';
	import { convertToUnit } from '../../helpers/convertToUnit';

	import { IndexedObject } from '../../types';

	import { mdiArrowUp } from '@mdi/js';

	import { Scroll } from 'vuetify/lib/directives';

	const Props = Vue.extend({
		props: {
			threshold: {
				type: Number,
				default: 120
			},
			nudgeRight: {
				type: [String, Number],
				default: '16px'
			},
			nudgeBottom: {
				type: [String, Number],
				default: '16px'
			},
			target: {
				type: String,
				default: undefined
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component({
		directives: {
			Scroll
		}
	})
	export default class BackToTopBtn extends MixinsDeclaration {
		locales = locales;

		topIcon = mdiArrowUp;

		showBtn = false;

		get targetSelector(): string | null {
			if (!this.target) {
				return null;
			}

			return `#${this.target}`;
		}

		get isMobile(): boolean {
			return this.$vuetify.breakpoint.smAndDown;
		}

		get btnStyle(): IndexedObject {
			const right = convertToUnit(this.nudgeRight) || '0';
			const bottom = convertToUnit(this.nudgeBottom) || '0';

			return {
				bottom,
				right
			};
		}

		get minWidth(): string | null {
			return this.isMobile ? '36px' : null;
		}

		get labelClasses(): IndexedObject<boolean> {
			return { 'd-sr-only': this.isMobile };
		}

		onScroll(e: MouseEvent): void {
			const target = e.currentTarget as HTMLElement | Window;

			if (target === window) {
				this.showBtn = window.scrollY > this.threshold;
			} else {
				this.showBtn = (target as HTMLElement).scrollTop > this.threshold;
			}
		}

		scrollToTop(): void {
			const target = document.getElementById(this.target) || window;
			target.scrollTo(0, 0);
		}
	}
</script>
