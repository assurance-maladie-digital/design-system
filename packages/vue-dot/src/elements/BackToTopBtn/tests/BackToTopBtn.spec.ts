import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import BackToTopBtn from '../';

let wrapper: Wrapper<Vue>;

describe('BackToTopBtn', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(BackToTopBtn);

		expect(wrapper).toMatchSnapshot();
	});
});
