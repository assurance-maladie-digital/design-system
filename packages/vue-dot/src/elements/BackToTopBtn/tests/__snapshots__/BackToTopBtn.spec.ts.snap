// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTopBtn renders correctly 1`] = `
<vfadetransition-stub origin="top center 0">
  <vbtn-stub color="primary" outlined="true" tag="button" activeclass="" fixed="true" type="button" class="vd-back-to-top-btn text-wrap px-0 px-md-4" style="bottom: 16px; right: 16px; display: none;"><span class="">
				Retour en haut
			</span>
    <vicon-stub color="primary" small="true" class="ml-0 ml-md-1">
      M13,20H11V8L5.5,13.5L4.08,12.08L12,4.16L19.92,12.08L18.5,13.5L13,8V20Z
    </vicon-stub>
  </vbtn-stub>
</vfadetransition-stub>
`;
